expr/plan.go

```diff
@@ -203,159 +117,99 @@ func newplanFunc(e *expr, fn Func, from, to uint32, stable bool, reqs []Req) ([]
 	//   might be dynamically typed. e.g. movingAvg returns 1..N series depending on how many it got as input
 
 	// first validate the mandatory args
-
-	j := 0 // pos in args of next given arg to process
-	for _, argExp := range argsExp {
-		if len(e.args) <= j {
+	pos := 0    // pos in args of next given arg to process
+	cutoff := 0 // marks the index of the first optional point (if any)
+	var argExp Arg
+	for cutoff, argExp = range argsExp {
+		if argExp.Optional() {
+			break
+		}
+		if len(e.args) <= pos {
 			return nil, ErrMissingArg
 		}
-		j, err = consumeArg(e.args, j, argExp)
+		pos, err = e.consumeBasicArg(pos, argExp)
```

Comment:
every time i read this code i get confused about how the values get assigned.  lets add comment that explains that the arg value will be assigned to argExp.Val, as it is not obvious